client/src/Auth/TabPanel.tsx

```diff
@@ -12,55 +13,16 @@ interface Props {
 
 const TabPanel: React.FC<Props> = ({ activeTab, index }) => {
     const classes = useStyles();
-
-    const [
-        { fullName, username, password, confirmPassword },
-        setCredentials,
-    ] = useState({
-        fullName: '',
-        username: '',
-        password: '',
-        confirmPassword: '',
-    });
-
-    const [error, setError] = useState('');
-
-    const LOGIN_TAB_INDEX = 1;
-    const PASSWORD_LENGTH_LIMIT = 4;
-
-    const isValidUsername = /^[0-9a-zA-Z_.-]+$/.test(username);
-    const isValidPassword = password.length > PASSWORD_LENGTH_LIMIT;
-    const isValidConfirmedPassword = confirmPassword === password;
-
-    const handleChange = ({
-        target: { name, value },
-    }: React.ChangeEvent<HTMLInputElement>) => {
-        setCredentials(c => ({ ...c, [name]: value }));
-    };
-
-    const register = async (event: React.FormEvent) => {
-        event.preventDefault();
-        if (!isValidUsername) {
-            setError('Username is invalid');
-        } else if (!isValidPassword) {
-            setError('Password is invalid');
-        } else if (!isValidConfirmedPassword) {
-            setError("Passwords don't match");
-        } else {
-            return null;
-        }
-    };
-
-    const login = async (event: React.FormEvent) => {
-        event.preventDefault();
-        if (!isValidUsername) {
-            setError('Username is invalid');
-        } else if (!isValidPassword) {
-            setError('Password is invalid');
-        } else {
-            return null;
-        }
-    };
+    const {
+        error,
+        userDetails: { fullName, username, password, confirmPassword },
+        isValidUsername,
+        isValidPassword,
+        isValidConfirmedPassword,
+        handleChange,
+        handleLogin,
+        handleRegister,
+    } = useTabPanel();
```

Comment:
Hmmm... Perhaps it is a good idea to uplift this hook to Auth.tsx file and pass props like "loginMode","ButtonText" and "SubmitCb" instead passing "activeTab" and "index" and then decide internally which action should be performed ?